service/history/visibility_queue_task_executor.go

```diff
@@ -370,44 +266,30 @@ func (t *visibilityQueueTaskExecutor) processCloseExecution(
 		return err
 	}
 
-	executionInfo := mutableState.GetExecutionInfo()
-	executionState := mutableState.GetExecutionState()
 	wfCloseTime, err := mutableState.GetWorkflowCloseTime(ctx)
 	if err != nil {
 		return err
 	}
-	workflowTypeName := executionInfo.WorkflowTypeName
-	workflowStatus := executionState.Status
-	workflowHistoryLength := mutableState.GetNextEventID() - 1
-	workflowStartTime := timestamp.TimeValue(mutableState.GetExecutionInfo().GetStartTime())
-	workflowExecutionTime := timestamp.TimeValue(mutableState.GetExecutionInfo().GetExecutionTime())
-	visibilityMemo := getWorkflowMemo(copyMemo(executionInfo.Memo))
-	searchAttr := getSearchAttributes(copySearchAttributes(executionInfo.SearchAttributes))
-	taskQueue := executionInfo.TaskQueue
+	historyLength := mutableState.GetNextEventID() - 1
+	executionInfo := mutableState.GetExecutionInfo()
 	stateTransitionCount := executionInfo.GetStateTransitionCount()
 	historySizeBytes := executionInfo.GetExecutionStats().GetHistorySize()
+	requestBase := t.getVisibilityRequestBase(task, namespaceEntry, mutableState)
 
 	// NOTE: do not access anything related mutable state after this lock release
 	// release the context lock since we no longer need mutable state and
 	// the rest of logic is making RPC call, which takes time.
 	release(nil)
-	err = t.recordCloseExecution(
+
+	err = t.visibilityMgr.RecordWorkflowExecutionClosed(
 		ctx,
-		namespaceEntry,
-		task.GetWorkflowID(),
-		task.GetRunID(),
-		workflowTypeName,
-		workflowStartTime,
-		workflowExecutionTime,
-		timestamp.TimeValue(wfCloseTime),
-		workflowStatus,
-		stateTransitionCount,
-		workflowHistoryLength,
-		task.GetTaskID(),
-		visibilityMemo,
-		taskQueue,
-		searchAttr,
-		historySizeBytes,
+		&manager.RecordWorkflowExecutionClosedRequest{
+			VisibilityRequestBase: requestBase,
+			CloseTime:             *wfCloseTime,
```

Comment:
Please make `wfCloseTime` a value (not pointer) and copy it's value before releasing the lock.